### PR TITLE
lib: updatehub: Add download block check

### LINKED ISSUE #25704: lib: updatehub: Corrupted updated when receiving CoAP duplicate packages
The updatehub library is not properly handling the case where the CoAP packages are duplicated and we end flashing a broken firmware on the device's storage.

The short term solution is to have an engine that handles these packages reception and ensures we handle the proper sequence of blocks.

### PR DESCRIPTION
**Pull request by nandojve**
The current implementation of UpdateHub uses stream to flash devices. That interface assume sequential data and the library requires that packets arrives on a sequential manner. Most wireless interfaces can suffer from package re transmission and out of order delivery that can cause wrong data block on flash. This patch sequence add CoAP block check to ensure the proper sequence. That avoids unnecessary downloads and useless flash write operations. A new mechanism to allow better re transmission requests, based on timers complement the CoAP block check. The combination of those techniques allow developer select proper configuration and tune re transmission mechanism allowing a fine tune by developer.

This work is based on #25558 contribution.

### CODE REVIEW

File: lib/updatehub/updatehub.c [hunk 1]
```diff
@@ -4,9 +4,11 @@
  * SPDX-License-Identifier: Apache-2.0
  */
 
-#include <logging/log.h>
+#define LOG_MODULE_NAME updatehub
+#define LOG_LEVEL CONFIG_UPDATEHUB_LOG_LEVEL
 
-LOG_MODULE_REGISTER(updatehub);
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
Comment (jukkar) [inline lib/updatehub/updatehub.c:12]: This works, but can be written shorter way (up to you if you want to change that).
```
#include <logging/log.h>
LOG_MODULE_REGISTER(updatehub, CONFIG_UPDATEHUB_LOG_LEVEL);
```

File: lib/updatehub/updatehub_timer.c [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * Copyright (c) 2020 O.S.Systems
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define LOG_MODULE_NAME updatehub_timer
+#define LOG_LEVEL CONFIG_UPDATEHUB_LOG_LEVEL
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(LOG_MODULE_NAME);
```
Comment (jukkar) [inline lib/updatehub/updatehub_timer.c:11]: We could utilize the same logging module here like this
```
#include <logging/log.h>
LOG_MODULE_DECLARE(updatehub, CONFIG_UPDATEHUB_LOG_LEVEL);
```

**Review by jukkar — APPROVED** [review]
LGTM

File: lib/updatehub/updatehub.c (hunk 1 above)
Comment (nandojve) [reply lib/updatehub/updatehub.c:12]: Sure! I'll change it as you suggested.

File: lib/updatehub/updatehub_timer.c (hunk 2 above)
Comment (nandojve) [reply lib/updatehub/updatehub_timer.c:11]: I'll send update.

**Comment by mtahirbutt** [comment]
> 
> 
> The current implementation of UpdateHub uses stream to flash devices. That interface assume sequential data and the library requires that packets arrives on a sequential manner. Most wireless interfaces can suffer from package re transmission and out of order delivery that can cause wrong data block on flash. This patch sequence add CoAP block check to ensure the proper sequence. That avoids unnecessary downloads and useless flash write operations. A new mechanism to allow better re transmission requests, based on timers complement the CoAP block check. The combination of those techniques allow developer select proper configuration and tune re transmission mechanism allowing a fine tune by developer.
> 
> This work is based on #25558 contribution.

also works fine for me. Can you increase maximum limit on CoAP retransmission (lets say 16-20) as wireless interfaces are problematic, besides right now if the attempt is unsuccessful, we have to reset the device in any case.

**Review by d3zd3z — APPROVED** [review]

**Pull request merged**